src/server/conn.rs

```diff
@@ -513,3 +553,165 @@ where
         }
     }
 }
+
+// TODO: different file? idk where to put this.
+mod rewindable {
+    use bytes::{Buf, BufMut, Bytes, IntoBuf};
+    use futures::{Async, Poll};
+    use std::io::{self, Read, Write};
+    use tokio_io::{AsyncRead, AsyncWrite};
+
+    #[derive(Debug)]
+    pub struct Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        pre: Option<Bytes>,
+        inner: T,
+    }
+
+    impl<T> Rewindable<T>
+    where
+        T: AsyncRead + AsyncWrite,
+    {
+        pub(super) fn new(tcp: T) -> Rewindable<T> {
+            Rewindable {
+                pre: None,
+                inner: tcp,
+            }
+        }
+        pub fn prepend(&mut self, bs: Bytes) {
+            self.pre.get_or_insert(bs);
```

Comment:
Prefer `self.pre = Some(bs)` instead, since `get_or_insert` does some matching on the original field, but we aren't using it.

You could put a `debug_assert!(self.pre.is_none())` before setting it, for correctness while tests are run.